controller/grpc.go

```diff
@@ -0,0 +1,1014 @@
+			}
+
+			return ctx, nil
+		}
+
+		return ctx, grpc.Errorf(codes.Unauthenticated, "no Auth-Key provided")
+	}
+
+	return ctx, grpc.Errorf(codes.Unauthenticated, "metadata missing")
+}
+
+func (g *grpcAPI) logger(ctx context.Context) log.Logger {
+	log, ok := ctxhelper.LoggerFromContext(ctx)
+	if !ok {
+		log = logger
+	}
+	return log
+}
+
+func (g *grpcAPI) logRequest(ctx context.Context, rpcMethod string) (context.Context, func(context.Context, error)) {
+	startTime := time.Now()
+
+	var clientIP string
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		var reqID []string
+		if id, ok := ctxhelper.RequestIDFromContext(ctx); ok {
+			reqID = []string{id}
+		} else {
+			reqID = md.Get("x-request-id")
+		}
+		if len(reqID) == 0 {
+			reqID = []string{random.UUID()}
+		}
+		ctx = ctxhelper.NewContextRequestID(ctx, reqID[0])
+		ctx = ctxhelper.NewContextLogger(ctx, g.logger(ctx).New("req_id", reqID[0]))
+
+		xForwardedFor := md.Get("x-forwarded-for")
+		if len(xForwardedFor) > 0 {
+			clientIPs := strings.Split(xForwardedFor[0], ",")
+			if len(clientIPs) > 0 {
+				clientIP = strings.TrimSpace(clientIPs[len(clientIPs)-1])
+			}
+		}
+	}
+	if clientIP == "" {
+		if remoteHost, ok := ctx.Value(ctxKeyRemoteHost).(string); ok {
+			clientIP = remoteHost
+		}
+	}
+
+	g.logger(ctx).Info("gRPC request started", "rpcMethod", rpcMethod, "client_ip", clientIP)
+	return ctx, func(ctx context.Context, err error) {
+		duration := time.Since(startTime)
+		if err == nil {
+			g.logger(ctx).Info("gRPC request ended", "duration", duration)
+		} else {
+			g.logger(ctx).Info("gRPC request ended", "duration", duration, "err", err)
+		}
+	}
+}
+
+type grpcEventListener struct {
+	Events  chan *ct.Event
+	Err     error
+	errOnce sync.Once
+	subs    []*data.EventSubscriber
+}
+
+func (e *grpcEventListener) Close() {
+	for _, sub := range e.subs {
+		sub.Close()
+		if err := sub.Err; err != nil {
+			e.errOnce.Do(func() { e.Err = err })
+		}
+	}
+}
+
+func (g *grpcAPI) subscribeEvents(appIDs []string, objectTypes []ct.EventType, objectIDs []string) (*grpcEventListener, error) {
+	dataEventListener, err := g.maybeStartEventListener()
+	if err != nil {
+		return nil, protobuf.NewError(err, err.Error())
+	}
+
+	eventListener := &grpcEventListener{
+		Events: make(chan *ct.Event),
+	}
+
+	objectTypeStrings := make([]string, len(objectTypes))
+	for i, t := range objectTypes {
+		objectTypeStrings[i] = string(t)
+	}
+
+	if len(appIDs) == 0 && len(objectIDs) == 0 {
+		// an empty string matches all app ids
+		appIDs = []string{""}
+	}
+	subs := make([]*data.EventSubscriber, 0, len(appIDs)+len(objectIDs))
+	for _, appID := range appIDs {
+		sub, err := dataEventListener.Subscribe(appID, objectTypeStrings, "")
+		if err != nil {
+			return nil, protobuf.NewError(err, err.Error())
+		}
+		subs = append(subs, sub)
+		go (func() {
+			for {
+				event, ok := <-sub.Events
+				if !ok {
+					break
+				}
+				eventListener.Events <- event
```

Comment:
This will block indefinitely and leak the goroutine if `eventListener` is closed (which indicates the subscriber is no longer reading from it), we should probably select on a stop channel of some sort